### DISCUSSION TITLE
Timer not as expected

### DISCUSSION THREAD

**Discussion by tkzcfc** [libuv/libuv#3660] [Q&A]
libuv: 1.19.2
platform: Windows

`

        static int fps = 0;
        auto loop = uv_default_loop();

        uv_timer_t timer1;
        uv_timer_init(loop, &timer1);
        uv_timer_start(&timer1, [](uv_timer_t* handle) {
                fps++;
        }, 1, 1);

        // fps timer
        uv_timer_t timer2;
        uv_timer_init(loop, &timer2);
        uv_timer_start(&timer2, [](uv_timer_t* handle) {
                printf("fps:%d\n", fps);
                fps = 0;
        }, 0, 1000);

        // loop
        uv_run(loop, UV_RUN_DEFAULT);
`

console output：
fps:0
fps:65
fps:64
fps:65
fps:65
fps:65
fps:65
fps:65

Why not output "fps:1000"
I need a more accurate timer

**Comment by vtjnash** [comment]
Use Unix then? Windows system timers may have roughly a 60fps cap (unless you are simultaneously running a media application that changes that internal default)

**Reply by tkzcfc** [reply]
Do you know how to break this limit

**Comment by tkzcfc** [comment]
I think it can be done on windows. When I add an idle, it works well, but the CPU load reaches 100%. I think the timeout time calculated by libuv is too long when it has only timer tasks

` 

	static int fps = 0;
	auto loop = uv_default_loop();

	uv_timer_t timer1;
	uv_timer_init(loop, &timer1);
	uv_timer_start(&timer1, [](uv_timer_t* handle) {
		fps++;
		}, 1, 1);

	uv_timer_t timer2;
	uv_timer_init(loop, &timer2);
	uv_timer_start(&timer2, [](uv_timer_t* handle) {
		printf("fps:%d\n", fps);
		fps = 0;
		}, 0, 1000);


	uv_idle_t idle;
	uv_idle_init(loop, &idle);
	uv_idle_start(&idle, [](uv_idle_t*) {
	});

	uv_run(loop, UV_RUN_DEFAULT);
`

output：
fps:0
fps:999
fps:1000
fps:1000
fps:1000
fps:1000
fps:1000
fps:1000
fps:1000
fps:1000

**Reply by vtjnash** [reply]
No, it is the timeout calculated by your operating system that adds to the latency

**Comment by tkzcfc** [comment]
OK, I tracked the code. When libuv called the pgetqueuedcompletionstatusex function, the incoming timeout was 1ms, but it worked for about 16ms
So to get an accurate timer, you need to make libuv busy

**Comment by vtjnash** [comment]
Correct. libuv has no trouble calculating the timeout correctly, but your OS normally does not support it

**Reply by tkzcfc** [reply]
OK, thank you for your answer. Can I break this limit

**Comment by tkzcfc** [comment]
I solved this problem using timeBeginPeriod 

https://randomascii.wordpress.com/2020/10/04/windows-timer-resolution-the-great-rule-change/
